### ISSUE TITLE
Support overriding the http transport applicationHeaderPrefix

### ISSUE THREAD

**Issue by mhamrah** [yarpc/yarpc-go#1261]
I want to copy all http headers in an incoming http request to the incoming context, however, the current http handler has

```
Headers:   applicationHeaders.FromHTTPHeaders(req.Header, 
```

with no way of modifying the applicationHeaders variable, which only copies headers with a prefix of ```ApplicationHeaderPrefix```.

I'd like to pass this prefix as an option new creating the inbound.

**Comment by breerly** [comment]
@mhamrah just to be clear - you'd like to read from a different header prefix, for example `MIKE-Procedure` instead of `RPC-Procedure` and so on? What about for outbound requests?

Is the background here that you're mapping to some existing HTTP dialect?

**Comment by mhamrah** [comment]
I specifically want to access the x-auth-user-email http header to identify who's calling the service internally when an http request is made. Open to alternatives in identifying the caller.

**Comment by breerly** [comment]
Guessing it's not possible to have them confirm with the transports header convention?

```
RPC-Header-Auth-User-Email
```

The above would end up in a `Auth-User-Email` header. The transport uses the prefix so that every single HTTP header isn't mapped as an RPC header.

**Comment by mhamrah** [comment]
No, it's injected via Uber's auth middleware.

**Comment by mhamrah** [comment]
I mean, I could possibly hack something in an upstream proxy, but would
really like the option to map all headers. Is there a reason this was
intentionally prevented?

**Comment by breerly** [comment]
Introducing variations in the wire format is roughly equivalent to introducing a new wire protocol, e.g. you might have variation in understanding between services.

I think it's quite possible to map custom headers into valid RPC headers. The bigger question is, can we uniformly do the same for all services? That way we can at least do this centrally, which should prevent a bunch of variation within the D.C.

**Comment by breerly** [comment]
@mhamrah can you schedule a 15 min chat w me? I think it'd help to understand the usecase just a little more.

**Comment by mhamrah** [comment]
Happy to chat. Also did some digging because we were using this header with
tchannel, and it doesn't appear the tchannel inbound filters headers the
same way the http inbound does. It just copies all incoming headers to the
context:
https://sourcegraph.com/github.com/yarpc/yarpc-go@2e024b59704f6511255a25628cedf6537a45d862/-/blob/transport/tchannel/header.go#L75-76

**Comment by abhinav** [comment]
@mhamrah The difference between TChannel and HTTP headers processing exists
because TChannel namespaces application headers to their own section.

HTTP headers include things like `Content-Type`, `Content-Length`, YARPC
headers (`Rpc-Service`, etc.). YARPC considers these transport-specific
details which should be abstracted away from the application layer. Having
access to raw HTTP headers breaks the guarantee that your application code is
agnostic of the transport being used.

That said, the HTTP inbound can probably expose an option to whitelist certain
HTTP headers to be passed-through as application headers. Something like,

```go
httpTransport.NewInbound(":0", http.GrabHeaders("X-Auth-Whatever", "X-Foo"))
// Function name TBD
```

Which would let you do,

```go
v := yarpc.CallFromContext(ctx).Header("X-Auth-Whatever")
```
    
We have precedent for this already. HTTP outbounds supports sending hard-coded
headers in requests using the [`AddHeader`][1] option.

[1]: https://godoc.org/go.uber.org/yarpc/transport/http#AddHeader

It's a slippery surface, though. It would be easy for people to start sticking
more and more information into the headers, completely negating one of the
main features of YARPC. To minimize leakage from the transport to the
application layer, perhaps GrabHeaders should only support `X-` headers in the
initial offering. If we see demand and a safe way to do it, we can open it up
to other header names.

How do other YARPC folks feel about this?

**Comment by breerly** [comment]
Ok, just chatted with @mhamrah, and I'd like us to support this so that we can retrofit our standard HTTP dialect with existing internal headers.

@abhinav lets sink and come up with a design that:

* Allows mapping to AND from HTTP headers to a separate header key in YARPC
* Allows the header key to match the one used in TChannel
* Avoids leaking transport-level details

I created a ticket internally to make sure we don't leave @mhamrah hangin'

**Comment by breerly** [comment]
Ok @mhamrah, jammed with @abhinav offline - I'm onboard with the design he suggested.

Does that look good to you?

**Comment by mhamrah** [comment]
Yes

**Comment by peter-edge** [comment]
@mhamrah this is merged and released in `v1.17.0`, `glide update` your project to get this.

**Comment by mhamrah** [comment]
Thanks!
